

> **Regression**

I will build a model to predict the **cost of treatment** for individuals based on their age, sex, bmi and other information.

Dataset description:




*   age: age of primary beneficiary

*   sex: insurance contractor gender, female, male

*   bmi: Body mass index, providing an understanding of body, weights that are relatively high or low relative to height, objective index of body weight (kg / m ^ 2) using the ratio of height to weight, ideally 18.5 to 24.9

*   children: Number of children covered by health insurance / Number of dependents

*   smoker: Smoking

*   region: the beneficiary's residential area in the US, northeast, southeast, southwest, northwest.

*   **charges: Individual medical costs billed by health insurance** (target)




#Importing Libraries

In [59]:
import numpy as np
import pandas as pd


#Importing Dataset

In [61]:
df=pd.read_csv("medical_cost.csv")
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

#Data Understanding

In [62]:
df.shape

(1338, 7)

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1335 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [64]:
df.describe()

age          bmi     children       charges
count  1338.000000  1335.000000  1338.000000   1338.000000
mean     39.207025    30.661423     1.094918  13270.422265
std      14.049960     6.101038     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.302500     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.687500     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

#Data Preprocessing

##Missing value

In [65]:
df.isnull().sum()

age         0
sex         0
bmi         3
children    0
smoker      0
region      0
charges     0
dtype: int64

We have only 3 missing values in bmi colomn compared to 1335 present values, so we can drop rows with these missing values because they don't affect the big number of data.
**Note**: if we would not drop rows, we can visualise the distribution of bmi to know by wich value (mean, mode, median) we should replace the missing one







In [66]:
df = df.dropna()
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [67]:
df.reset_index(drop=True, inplace=True)

In [68]:
df.shape

(1335, 7)

##Categorical features

In [69]:
df=pd.get_dummies(df).drop(['sex_female','smoker_no','region_southwest'], axis=1)
df.head(5)

age     bmi  children      charges  sex_male  smoker_yes  region_northeast  \
0   19  27.900         0  16884.92400         0           1                 0   
1   18  33.770         1   1725.55230         1           0                 0   
2   28  33.000         3   4449.46200         1           0                 0   
3   33  22.705         0  21984.47061         1           0                 0   
4   32  28.880         0   3866.85520         1           0                 0   

   region_northwest  region_southeast  
0                 0                 0  
1                 0                 1  
2                 0                 1  
3                 1                 0  
4                 1                 0

In [70]:
df.shape

(1335, 9)

**Note**: Values in my dataset are not very far from each other, so i would not do the normalization of values

#Data Spliting

In [71]:
x = df[['age','bmi','children','sex_male','smoker_yes','region_northeast','region_northwest','region_southeast']]
y = df.iloc[:, 3]


In [72]:
x.head()

age     bmi  children  sex_male  smoker_yes  region_northeast  \
0   19  27.900         0         0           1                 0   
1   18  33.770         1         1           0                 0   
2   28  33.000         3         1           0                 0   
3   33  22.705         0         1           0                 0   
4   32  28.880         0         1           0                 0   

   region_northwest  region_southeast  
0                 0                 0  
1                 0                 1  
2                 0                 1  
3                 1                 0  
4                 1                 0

In [73]:
x.shape

(1335, 8)

In [74]:
y

0       16884.92400
1        1725.55230
2        4449.46200
3       21984.47061
4        3866.85520
           ...     
1330    10600.54830
1331     2205.98080
1332     1629.83350
1333     2007.94500
1334    29141.36030
Name: charges, Length: 1335, dtype: float64

In [75]:
y.shape

(1335,)

In [76]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

#Linear Regression model

In [77]:
from sklearn.linear_model import LinearRegression
r1 = LinearRegression()
r1.fit(x_train,y_train)

LinearRegression()

##Prediction

In [78]:
y_pred = r1.predict(x_test)

In [79]:
y_test

1016    12495.29085
445      7345.08400
458     10601.41200
251     47305.30500
980      4500.33925
           ...     
805     28476.73499
883      4877.98105
194      1137.46970
240     38511.62830
1165     4670.64000
Name: charges, Length: 267, dtype: float64

In [80]:
y_pred.reshape(-1,1)

array([[ 1.46844700e+04],
       [ 9.47931092e+03],
       [ 1.45721296e+04],
       [ 3.84342465e+04],
       [ 4.35807528e+03],
       [ 1.00384594e+04],
       [ 3.79013856e+04],
       [ 4.11858411e+03],
       [ 4.60728896e+03],
       [ 3.62072733e+04],
       [ 9.41618788e+03],
       [ 3.00298469e+04],
       [ 1.04310849e+04],
       [ 2.69830607e+03],
       [ 7.30364861e+03],
       [ 7.26471411e+02],
       [ 1.43600357e+04],
       [ 2.92152327e+04],
       [ 1.13106483e+04],
       [ 3.30018277e+04],
       [ 3.10714596e+03],
       [ 4.65618368e+03],
       [ 5.23996999e+03],
       [ 3.72955308e+03],
       [ 4.14491485e+03],
       [ 1.36024489e+04],
       [ 1.52143547e+03],
       [-2.16839277e+00],
       [ 3.31832705e+04],
       [ 2.90524716e+03],
       [ 2.73858327e+04],
       [ 8.87473855e+03],
       [ 1.26367179e+04],
       [ 3.93570007e+04],
       [ 9.34986475e+03],
       [-1.08246187e+03],
       [ 4.47667704e+03],
       [ 8.47712240e+03],
       [ 7.4

##Evaluation

In [81]:
from sklearn.metrics import r2_score
print(r2_score(y_test, y_pred))

0.7578912886388554


#Decision Tree

In [82]:
from sklearn.tree import DecisionTreeRegressor
r2= DecisionTreeRegressor()
r2.fit(x_train,y_train)

DecisionTreeRegressor()

##Prediction

In [83]:
y_pred_decision_tree = r2.predict(x_test)

In [84]:
y_test

1016    12495.29085
445      7345.08400
458     10601.41200
251     47305.30500
980      4500.33925
           ...     
805     28476.73499
883      4877.98105
194      1137.46970
240     38511.62830
1165     4670.64000
Name: charges, Length: 267, dtype: float64

In [85]:
y_pred_decision_tree.reshape(-1,1)

array([[12269.68865 ],
       [17929.30337 ],
       [10982.5013  ],
       [46130.5265  ],
       [21984.47061 ],
       [24915.04626 ],
       [44202.6536  ],
       [ 2473.3341  ],
       [ 6402.29135 ],
       [43254.41795 ],
       [ 7789.635   ],
       [21259.37795 ],
       [ 7518.02535 ],
       [21344.8467  ],
       [ 6128.79745 ],
       [ 2597.779   ],
       [12231.6136  ],
       [21472.4788  ],
       [10796.35025 ],
       [24915.22085 ],
       [ 2020.177   ],
       [ 7133.9025  ],
       [ 3172.018   ],
       [ 3757.8448  ],
       [ 2464.6188  ],
       [10226.2842  ],
       [ 2196.4732  ],
       [ 1731.677   ],
       [38709.176   ],
       [ 4719.73655 ],
       [36307.7983  ],
       [ 7682.67    ],
       [25333.33284 ],
       [49577.6624  ],
       [ 7445.918   ],
       [ 1242.26    ],
       [ 1526.312   ],
       [ 3935.1799  ],
       [ 5594.8455  ],
       [10096.97    ],
       [ 9193.8385  ],
       [ 8219.2039  ],
       [ 4149.736   ],
       [ 56

##Evaluation

In [86]:
print(r2_score(y_test, y_pred_decision_tree))

0.6877368722406159


#Random Forest

In [87]:
from sklearn.ensemble import RandomForestRegressor
r3 = RandomForestRegressor(n_estimators=150)
r3.fit(x_train, y_train)

RandomForestRegressor(n_estimators=150)

##Precdiction

In [88]:
y_pred_random_forest = r3.predict(x_test)


In [89]:
y_test

1016    12495.29085
445      7345.08400
458     10601.41200
251     47305.30500
980      4500.33925
           ...     
805     28476.73499
883      4877.98105
194      1137.46970
240     38511.62830
1165     4670.64000
Name: charges, Length: 267, dtype: float64

In [90]:
y_pred_random_forest.reshape(-1,1)

array([[16923.7977188 ],
       [ 9547.5591742 ],
       [11650.91015853],
       [46619.95626913],
       [ 7531.2827316 ],
       [13569.4157794 ],
       [43965.460741  ],
       [ 3587.16078607],
       [ 8165.43129727],
       [44275.8904554 ],
       [ 8641.9804472 ],
       [25924.1202462 ],
       [ 8475.63139173],
       [14452.97124667],
       [ 5796.72035907],
       [ 2360.07021353],
       [12190.57348227],
       [19641.52899067],
       [12371.15616767],
       [25302.26901727],
       [ 3945.90620827],
       [ 6935.0774166 ],
       [ 4323.37920767],
       [ 3868.65272593],
       [ 3556.6498964 ],
       [11711.38823873],
       [ 2207.91455667],
       [ 1968.26248633],
       [39288.62383133],
       [ 5965.46996847],
       [36425.13056633],
       [ 7719.66084733],
       [12427.9655974 ],
       [47060.42559607],
       [ 8239.52650607],
       [ 1306.070011  ],
       [ 3477.05282693],
       [ 7844.1284036 ],
       [ 6785.5215598 ],
       [10175.30292773],


##Evaluation

In [91]:
print(r2_score(y_test, y_pred_random_forest))

0.8387759967094282


#XGBoost

In [92]:
from xgboost import XGBRegressor
r4 = XGBRegressor()
r4.fit(x_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

##Prediction

In [93]:
y_pred_xgboost = r4.predict(x_test)


In [94]:
y_test

1016    12495.29085
445      7345.08400
458     10601.41200
251     47305.30500
980      4500.33925
           ...     
805     28476.73499
883      4877.98105
194      1137.46970
240     38511.62830
1165     4670.64000
Name: charges, Length: 267, dtype: float64

In [95]:
y_pred_xgboost.reshape(-1,1)

array([[17275.625  ],
       [ 9829.545  ],
       [11771.396  ],
       [46219.574  ],
       [ 7240.169  ],
       [ 8195.927  ],
       [46882.49   ],
       [ 5544.8745 ],
       [11736.306  ],
       [42237.957  ],
       [ 8312.035  ],
       [24503.844  ],
       [10203.091  ],
       [18561.674  ],
       [ 8489.149  ],
       [ 1080.9222 ],
       [11098.462  ],
       [21244.502  ],
       [12687.16   ],
       [25409.697  ],
       [ 1980.711  ],
       [ 8063.349  ],
       [ 5428.7837 ],
       [ 4181.676  ],
       [ 3630.8472 ],
       [13376.534  ],
       [ 1816.7288 ],
       [  725.9705 ],
       [38145.242  ],
       [ 4656.7314 ],
       [36061.52   ],
       [ 7453.8613 ],
       [12366.632  ],
       [49432.08   ],
       [11655.945  ],
       [ 1222.3256 ],
       [ 2652.4688 ],
       [ 5544.1504 ],
       [ 7242.428  ],
       [10247.899  ],
       [ 8986.967  ],
       [ 9227.218  ],
       [10245.643  ],
       [ 9157.141  ],
       [10694.384  ],
       [37

##Evaluation

In [96]:
print(r2_score(y_test, y_pred_xgboost))

0.8221803767859728


#Conclusion

After doing thes 4 models for medical_cost dataset we can see that:


*   Linear Regression Model has a R2 score: 0.7578912886388554

*   Decision Tree Model has a R2 score: 0.6975601083504588

*   Random Forest Model has a R2 score: 0.8417671054977912

*   XGBoost Model has a R2 score: 0.8221803767859728

So we can conclude that the best model to be adapted for a great prediction of the medical cost is the Model that have the nearest R2 score to 1. So the best model is **Random Forest**




**Note**: i used only the R2 score in my evaluation because it is the best factor for  a great evaluation.